src/mds/Locker.cc

```diff
@@ -225,6 +225,35 @@ struct MarkEventOnDestruct {
   }
 };
 
+/*
+ * Try to detect the possible deadlock between requests, and
+ * will try to break the newer request.
+ */
+void Locker::detect_dead_locks(const MDRequestRef& mdr, int lock_type)
+{
+  bool any_waiter = false;
+  bool order_mess = false;
+
+  dout(15) << "detect_dead_locks " << *mdr << dendl;
+  for (auto it = mdr->locks.begin(); it != mdr->locks.end(); ) {
+    SimpleLock *lock = it->lock;
+
+    if (lock->is_waiter_for(SimpleLock::WAIT_ALL)) {
+      any_waiter = true;
+    }
+    if (lock->get_type() > lock_type) {
+      order_mess = true;
+    }
+    dout(15) << " lock " << *lock << " any_waiter " << any_waiter << " order_mess " << order_mess << dendl;
+  }
+
+  if (order_mess && any_waiter) {
+    dout(10) << " detected possible dead locks; dropping all locks" << dendl;
+    drop_locks(mdr.get(), NULL);
+    mdr->drop_local_auth_pins();
```

Comment:
If we chose to drop locks on a request then we _must_ terminate this `acquire_locks` call with a `false` and put the request on the replay queue without any timeout

Reply:
For this the `rdlock_start(), etc` has already done it. IMO we still need to wait the corresponding locks to be ready in `rdlock_start(), etc`, else this request could be fired requently in a loop by doing this check here.

Reply:
> For this the `rdlock_start(), etc` has already done it.

Yes, you are right, the request is already on a waiting list. But it's on a waiting list for a wrong lock. We need to re-run the request from scratch now so that it will restart acquiring locks in the correct order, waiting on the appropriate locks this time.